File: packages/core/src/parse.ts
import type { ConflictHunk, ParsedConflicts } from "./types.ts";

const MARKER = /^<{7}( |$)|^={7}$|^>{7}( |$)|^\|{7}( |$)/m;
// Git emits exactly 7 marker chars; longer runs (setext underlines, ASCII
// rules) inside conflict content must not read as markers.
const OPEN = /^<{7}( |$)/;
const BASE = /^\|{7}( |$)/;
const SEP = /^={7}$/;
const CLOSE = /^>{7}( |$)/;

export function hasConflictMarkers(text: string): boolean {
  return MARKER.test(text);
}

export class ConflictParseError extends Error {
  constructor(
    message: string,
    readonly line: number,
  ) {
    super(message);
    this.name = "ConflictParseError";
  }
}

function label(line: string): string {
  const i = line.indexOf(" ");
  return i === -1 ? "" : line.slice(i + 1).trim();
}

/**
 * Parse git conflict markers (two-way `<<<<<<<`/`=======`/`>>>>>>>` and
 * diff3 `|||||||` base sections). Throws ConflictParseError on unbalanced
 * markers — callers should treat that as escalate-everything.
 */
export function parseConflicts(text: string): ParsedConflicts {
  const lines = text.split("\n");
  const hunks: ConflictHunk[] = [];

  for (let i = 0; i < lines.length; i++) {
    if (!OPEN.test(lines[i])) continue;

    const hunk: ConflictHunk = {
      startLine: i,
      endLine: -1,
      oursLabel: label(lines[i]),
      theirsLabel: "",
      ours: [],
      base: null,
      theirs: [],
    };

    // section: 0 = ours, 1 = base, 2 = theirs
    let section = 0;
    let closed = false;
    for (let j = i + 1; j < lines.length; j++) {
      const l = lines[j];
      if (OPEN.test(l)) {
        throw new ConflictParseError("nested conflict marker", j);
      }
      if (BASE.test(l)) {
        if (section !== 0) throw new ConflictParseError("base marker out of place", j);
        section = 1;
        hunk.base = [];
        continue;
      }
      if (SEP.test(l)) {
        if (section === 2) throw new ConflictParseError("duplicate separator", j);
        section = 2;
        continue;
      }
      if (CLOSE.test(l)) {
        if (section !== 2) throw new ConflictParseError("no separator in conflict", j);
        hunk.theirsLabel = label(l);
        hunk.endLine = j + 1;
        closed = true;
        i = j;
        break;
      }
      if (section === 0) hunk.ours.push(l);
      else if (section === 1) hunk.base!.push(l);
      else hunk.theirs.push(l);
    }
    if (!closed) throw new ConflictParseError("unterminated conflict", hunk.startLine);
    hunks.push(hunk);
  }

  return { lines, hunks };
}

File: packages/core-go/parse.go
package core

import (
	"fmt"
	"regexp"
	"strings"
)

// Git emits exactly 7 marker chars; longer runs (setext underlines, ASCII
// rules) inside conflict content must not read as markers.
var (
	markerRe = regexp.MustCompile(`(?m)^<{7}( |$)|^={7}$|^>{7}( |$)|^\|{7}( |$)`)
	openRe   = regexp.MustCompile(`^<{7}( |$)`)
	baseRe   = regexp.MustCompile(`^\|{7}( |$)`)
	sepRe    = regexp.MustCompile(`^={7}$`)
	closeRe  = regexp.MustCompile(`^>{7}( |$)`)
)

func HasConflictMarkers(text string) bool { return markerRe.MatchString(text) }

// ConflictParseError reports unbalanced markers; callers should treat it as
// escalate-everything.
type ConflictParseError struct {
	Line int
	Msg  string
}

func (e *ConflictParseError) Error() string {
	return fmt.Sprintf("%s (line %d)", e.Msg, e.Line+1)
}

func markerLabel(line string) string {
	i := strings.IndexByte(line, ' ')
	if i == -1 {
		return ""
	}
	return strings.TrimSpace(line[i+1:])
}

// ParseConflicts parses two-way and diff3 (||||||| base section) markers.
func ParseConflicts(text string) (*ParsedConflicts, error) {
	lines := strings.Split(text, "\n")
	var hunks []ConflictHunk

	for i := 0; i < len(lines); i++ {
		if !openRe.MatchString(lines[i]) {
			continue
		}
		hunk := ConflictHunk{StartLine: i, EndLine: -1, OursLabel: markerLabel(lines[i])}
		section := 0 // 0 = ours, 1 = base, 2 = theirs
		closed := false
		for j := i + 1; j < len(lines); j++ {
			l := lines[j]
			if openRe.MatchString(l) {
				return nil, &ConflictParseError{Line: j, Msg: "nested conflict marker"}
			}
			if baseRe.MatchString(l) {
				if section != 0 {
					return nil, &ConflictParseError{Line: j, Msg: "base marker out of place"}
				}
				section = 1
				hunk.Base = []string{}
				continue
			}
			if sepRe.MatchString(l) {
				if section == 2 {
					return nil, &ConflictParseError{Line: j, Msg: "duplicate separator"}
				}
				section = 2
				continue
			}
			if closeRe.MatchString(l) {
				if section != 2 {
					return nil, &ConflictParseError{Line: j, Msg: "no separator in conflict"}
				}
				hunk.TheirsLabel = markerLabel(l)
				hunk.EndLine = j + 1
				closed = true
				i = j
				break
			}
			switch section {
			case 0:
				hunk.Ours = append(hunk.Ours, l)
			case 1:
				hunk.Base = append(hunk.Base, l)
			default:
				hunk.Theirs = append(hunk.Theirs, l)
			}
		}
		if !closed {
			return nil, &ConflictParseError{Line: hunk.StartLine, Msg: "unterminated conflict"}
		}
		hunks = append(hunks, hunk)
	}
	return &ParsedConflicts{Lines: lines, Hunks: hunks}, nil
}
